Record mode

darkdarkdragon opened the pull request:
**What does this pull request do? Explain your changes. (required)**
Allows to record streams 


**Specific updates (required)**
- add generic URL-based OS syntax https://github.com/livepeer/go-livepeer/pull/1582/commits/6b7b0c75280a300e906fe6f0cffeca3b906b0c20
- add `-objectStore` command line argument https://github.com/livepeer/go-livepeer/pull/1582/commits/4cfedeae2e46588f117c8ffe4b588c856493b4ec
- allow to specify OS per-stream in webhook https://github.com/livepeer/go-livepeer/pull/1582/commits/f768a952f4fef93d8d19e3ee0ed3966b3932d175

PR allows to specify global OS for saving streams (using `-recordStore` command line option) or to specify OS per-stream using webhook. 
Two fields added to webhook response structure:
`objectStore` and `recordObjectStore`. 
`objectStore` specifies URL of OS to use for stream in transcoding pipeline, and
`recordObjectStore` specifies URL of OS to store stream to.

Recorded streams can be played back using http://broadcaster:8935/recordings/manifestID/index.m3u8 URL.




**How did you test each of these updates (required)**
Units tests

**Does this pull request close any open issues?**
Fixes #1609 
Fixes #1572 


**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

iameli committed:
initial -record implementation

iameli committed:
drivers: add s3 url parsing

iameli committed:
drivers: add google store syntax

iameli committed:
cmd: add -objectStore, deprecate others

iameli committed:
drivers: add support for s3+http:// stores

iameli committed:
server: implement webhook-defined os

iameli committed:
drivers: implement support for gs ?keyfile=

iameli committed:
drivers: refactor IsOwnExternal to IsOwn

darkdarkdragon committed:
go mod update

iameli reviewed (dismissed):
So we're going to run into some trouble with `nodeId`.

1. It's not set unless `-monitor=true` is set, which results in an empty string being used in the S3 upload URLs and playback breaking.
2. In on-chain mode it uses the broadcaster address... except that'll be the same across all of our production broadcasters in a region, so their JSON uploads will all collide.

I dunno what we wanna do about that exactly.

iameli commented:
I tried to make a copy of this branch and accidentally deleted and recreated it, sorry if that causes problems anywhere 😂

darkdarkdragon committed:
Always initialise NodeID

darkdarkdragon committed:
Use AuthWebHook to find record object store for stream in `/recordings` handler

darkdarkdragon committed:
Better error reporting from auth webhook

darkdarkdragon committed:
Construct full URl to pass to webhook in /recordings endpoint

darkdarkdragon committed:
Cache AuthWebhook responses in /recordings handler

darkdarkdragon committed:
Log content of 'Content-Duration' header in the HTTP push handler

darkdarkdragon committed:
in HTTP push handler ensure that POST is used

darkdarkdragon committed:
Also allow PUT method for HTTP push handler

darkdarkdragon committed:
Output more debugging info in /recording handler

darkdarkdragon committed:
Add 10 seconds timeout to S3/GS driver's SaveData method

darkdarkdragon committed:
Retry saving segment to recordings OS once

darkdarkdragon committed:
Output more logging information

darkdarkdragon committed:
Read json playlists in parallel

darkdarkdragon committed:
Save JSON playlist to object storage in own goroutine
Fixes #1609

darkdarkdragon committed:
Prevent memory leak

darkdarkdragon committed:
Add unit tests for recordings

darkdarkdragon committed:
mod: Update LPMS to https://github.com/livepeer/lpms/commit/d5c85d8

darkdarkdragon committed:
Use `drivers.ParseOSURL` in tests
Add stub auth token to test

darkdarkdragon committed:
Save source stream's playlist if transcoding not happening

darkdarkdragon committed:
- use external manifest id for recording
 - add `recordObjectStoreUrl` field to the webhook auth response - used
  as absolute URL to the segments saved to the recording
  object storage

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -720,16 +722,53 @@ func main() {
+			Host:     *gsBucket,
+			RawQuery: fmt.Sprintf("keyfile=%s", *gsKey),
+		}
+		glog.Warningf("-gsbucket and -gskey are deprecated. Instead, you can use -objectStore '%s'", u.String())
```
Looks like extraneous single quotes around `%s`.

darkdarkdragon replied:
fixed

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -720,16 +722,53 @@ func main() {
+			glog.Error("Error creating object store driver: ", err)
+			return
+		}
+		drivers.NodeStorage, err = drivers.ParseOSURL(prepared, false)
```
Missing error handling.

darkdarkdragon replied:
Fixed

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -720,16 +722,53 @@ func main() {
+			glog.Error("Error creating recordings object store driver: ", err)
+			return
+		}
+		drivers.RecordStorage, err = drivers.ParseOSURL(prepared, true)
```
Missing error handling.

darkdarkdragon replied:
Fixed.

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -720,16 +722,53 @@ func main() {
 	if *gsBucket != "" && *gsKey != "" {
-		drivers.GSBUCKET = *gsBucket
-		drivers.NodeStorage, err = drivers.NewGoogleDriver(*gsBucket, *gsKey)
+		store, err := drivers.NewGoogleDriver(*gsBucket, *gsKey, false)
```
Do we still need the logic in this conditional if we set `drivers.NodeStorage` with the `if *objectstore != ""` conditional above?

darkdarkdragon replied:
Removed.

yondonfu commented on drivers/drivers.go:
```diff
@@ -46,14 +92,86 @@ func NewSession(info *net.OSInfo) OSSession {
+		return "", err
+	}
+	if u.Scheme == "gs" {
+		m, _ := url.ParseQuery(u.RawQuery)
```
Needs error handling.

darkdarkdragon replied:
Fixed.

yondonfu commented on core/playlistmanager.go:
```diff
@@ -19,35 +25,175 @@ type PlaylistManager interface {
+
+type JsonPlaylist struct {
+	name       string
+	DurationMs uint64               `json:"duration_ms,omitempty"` // total duration of the saved sagments
```
Nit: Match capitalization for this field in `jsonSeg` and here?

darkdarkdragon replied:
Fixed.

yondonfu commented on core/playlistmanager.go:
```diff
@@ -19,35 +25,175 @@ type PlaylistManager interface {
+		mindex := strings.Index(uri, manifestID)
+		if mindex != -1 {
+			if extURL != "" {
+				uri = path.Join(extURL, uri[mindex:])
```
a) `<extURL>/<manifestID>/...`

b) `/...`

So we expect the URI to be a) when there is an external URL and b) when there is not an external URL? Why do we include the manifestID in a) but exclude it in b)?

darkdarkdragon replied:
added as comment:
```
			// If extURL was specified we will put absolute URL to the segment into manifest
			// extURL points to the root of object store, so we should take part of the 'uri'
			// which contains manifestID.
			// If extURL is not specified then we're serving relative URL to the segment,
			// and address at which manifest is served already contains manfiestID
			// (broadcaster.com/recordings/manifestID/index.m3u8), so we're taking
			// part of the 'uri' after the manifestID
```

yondonfu commented on core/playlistmanager.go:
```diff
@@ -19,35 +25,175 @@ type PlaylistManager interface {
+}
+
+func (jpl *JsonPlaylist) hasTrack(trackName string) bool {
+	for _, track := range jpl.Tracks {
```
Note to self that this linear search should be ok because in practice the # of tracks = # of renditions which will typically be small (i.e. less than 10).

yondonfu commented on core/playlistmanager.go:
```diff
@@ -63,6 +209,31 @@ func (mgr *BasicPlaylistManager) GetOSSession() drivers.OSSession {
+			glog.V(common.VERBOSE).Infof("Saving json playlist name=%s size=%d bytes took=%s err=%v", name,
+				len(b), time.Since(now), err)
+		}(mgr.jsonList.name, b)
+		if mgr.jsonList.DurationMs > 60*60*1000 { // 1 hour
```
Use a constant here?

darkdarkdragon replied:
Fixed.

yondonfu commented on install_ffmpeg.sh:
```diff
@@ -115,7 +115,7 @@ if [ ! -e "$HOME/ffmpeg/libavcodec/libavcodec.a" ]; then
     --disable-postproc --disable-lzma \
     --enable-gnutls --enable-libx264 --enable-gpl \
-    --enable-protocol=https,rtmp,file \
+    --enable-protocol=https,http,rtmp,file \
```
Hm does this mean that http:// URLs couldn't be passed as filenames for transcoding previously? And is that the main reason for this change?

darkdarkdragon replied:
Honestly I don't remember, but probably we was using local min.io for OS - which is on http, of course.

yondonfu commented on server/broadcast.go:
```diff
@@ -299,7 +300,25 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) ([]string, erro
+	segDurMs := strconv.Itoa(int(seg.Duration * 1000))
+	if ros != nil {
+		go func() {
+			for i := 0; i < 2; i++ {
```
Can this logic be moved into a function since it needs to be executed for both the source as well as for each rendition?

darkdarkdragon replied:
Fixed.

yondonfu replied:
I was actually referring to the for loop with the retry code for `SaveData()` that is executed for the source as well as for each rendition.

darkdarkdragon replied:
fixed

yondonfu commented on server/broadcast.go:
```diff
@@ -299,7 +300,25 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) ([]string, erro
+					glog.Errorf("Error saving nonce=%d manifestID=%s name=%s to record store try=%d err=%v", nonce, mid, name, i, err)
+				} else {
+					cpl.InsertHLSSegmentJSON(vProfile, seg.SeqNo, uri, seg.Duration)
+					glog.Infof("Successfully saved nonce=%d manifestID=%s name=%s size=%d bytes to record store took=%s",
```
`size=` -> `bytes=` since we use that convention elsewhere? Also avoids the need to specify the units afterwords.

darkdarkdragon replied:
Fixed.

yondonfu commented on core/playlistmanager.go:
```diff
@@ -19,35 +25,175 @@ type PlaylistManager interface {
+				// x is not present in data,
+				// but i is the index where it would be inserted.
+				if i < len(curSegs) {
+					// glog.Errorf("track %s cur segs len %d i %d seq %d", trackName, len(curSegs), i, seg.SeqNo)
```
Let's clean this up.

darkdarkdragon replied:
Fixed.

yondonfu commented on drivers/drivers.go:
```diff
@@ -46,14 +92,86 @@ func NewSession(info *net.OSInfo) OSSession {
+}
+
+// ParseOSURL returns the correct OS for a given OS url
+func ParseOSURL(input string, useFullAPI bool) (OSDriver, error) {
```
When is useFullAPI supposed to be true vs. false?

darkdarkdragon replied:
false for -objectStore - OS used in transcoding pipeline, true for -recordStore.
This is because of objectStore we need to be able to pass OS credentials between broadcaster and orchestrator and we only need to be able to write segment and to read it back by name.
For recordStore we don't need to pass credentials, but we need to be able list files/dirs, write/read metadata etc.

yondonfu commented on drivers/gs.go:
```diff
@@ -95,21 +103,35 @@ func NewGoogleDriver(bucket, keyFileName string) (OSDriver, error) {
 		parsedKey: parsedKey,
 	}
+	if useFullAPI {
+		client, err := storage.NewClient(context.Background(), option.WithCredentialsJSON(os.keyData))
```
Looks like we're just creating a client and then immediately closing it. What is the purpose of this?

darkdarkdragon replied:
Probably I was trying to check if credentials are valid, but in reality it doesn't do anything useful. Removed.

yondonfu commented on drivers/gs.go:
```diff
@@ -125,6 +147,184 @@ func newGSSession(info *net.S3OSInfo) OSSession {
+func (os *gsSession) createClient() error {
+	client, err := storage.NewClient(context.Background(), option.WithCredentialsJSON(os.keyData))
+	if err != nil {
+		glog.Errorf("Error creating GCP client err=%v", err)
```
If we're returning the error let's let the caller handle it to avoid accidentally handling it in two places?

darkdarkdragon replied:
Fixed.

yondonfu commented on drivers/gs.go:
```diff
@@ -125,6 +147,184 @@ func newGSSession(info *net.S3OSInfo) OSSession {
+	}
+	if os.client == nil {
+		if err := os.createClient(); err != nil {
+			return nil, err
```
Why don't we just create and reuse a single client and store it when creating the session?

darkdarkdragon replied:
Because `NewSession` doesn't return error and so we don't have a way to surface error in case client creation fails.

yondonfu commented on drivers/local.go:
```diff
@@ -185,3 +257,21 @@ func (dc *dataCache) GetData(name string) []byte {
 	return nil
 }
+
+type dumbPageInfo struct {
```
Is a more descriptive name here something like `singlePageInfo` to convey that there is no next page?

darkdarkdragon replied:
Fixed.

yondonfu commented on server/broadcast.go:
```diff
@@ -525,7 +567,7 @@ func transcodeSegment(cxn *rtmpConnection, seg *stream.HLSSegment, name string,
 	if verifier != nil {
 		// verify potentially can change content of segURLs
-		err := verify(verifier, cxn, sess, seg, res.TranscodeData, segURLs, segData)
+		err := verify(verifier, cxn, sess, seg, res.TranscodeData, segURLs, segData, cpl.GetOSSession())
```
Passing cpl.GetOSSession() shouldn't be necessary since we're already passing in cxn which has access to the OSSession vis cxn.pl.GetOSSession()

darkdarkdragon replied:
Fixed.

yondonfu commented on server/rpc.go:
```diff
@@ -283,8 +283,12 @@ func orchestratorInfo(orch Orchestrator, addr ethcommon.Address, serviceURI stri
+		if os.IsExternal() {
+			tr.Storage = []*net.OSInfo{os.GetInfo()}
+		} else {
+			os.EndSession()
```
Why are we ending the session here?

darkdarkdragon replied:
Because in this branch session object is not used, but it will remain in the session's map
https://github.com/livepeer/go-livepeer/blob/72eea36e9cca167c5a565b2384fc7cd2e3c1e533/drivers/local.go#L49
so it will be a memory leak.

yondonfu commented on verification/verify.go:
```diff
@@ -57,6 +58,9 @@ type Params struct {
 	Renditions [][]byte
+
+	// External object storage used by broadcaster (can be defined per-stream)
+	OS drivers.OSSession
```
Why do we need this? URIs should already contain locations of the renditions when data is in an external OS.

darkdarkdragon replied:
But in verifier we need to distinguish between local OS and external OS

https://github.com/livepeer/go-livepeer/blob/0e3578d322ec397926f0eea9c63444ecf8e2a4e5/verification/epic.go#L191

yondonfu replied:
Ah so we removed drivers.IsOwnExternal and now we call IsExternal() and IsOwn() on each driver session.

yondonfu commented on server/mediaserver.go:
```diff
@@ -285,7 +326,7 @@ func authenticateStream(url string) (*authWebhookResponse, error) {
 	resp.Body.Close()
 	if resp.StatusCode != 200 {
-		return nil, errors.New(resp.Status)
+		return nil, fmt.Errorf("Status=%d error=%s", resp.StatusCode, string(rbody))
```
Nit: "Status" -> "status" to match capitalization of error

darkdarkdragon replied:
Fixed.

yondonfu commented on server/mediaserver.go:
```diff
@@ -648,6 +690,10 @@ func getRTMPStreamHandler(s *LivepeerServer) func(url *url.URL) (stream.RTMPVide
 func (s *LivepeerServer) HandlePush(w http.ResponseWriter, r *http.Request) {
 	start := time.Now()
+	if r.Method != "POST" && r.Method != "PUT" {
+		w.WriteHeader(http.StatusMethodNotAllowed)
```
Add an error log here?

darkdarkdragon replied:
Fixed.

yondonfu commented on server/mediaserver.go:
```diff
@@ -896,6 +949,254 @@ func (s *LivepeerServer) HandlePush(w http.ResponseWriter, r *http.Request) {
+// HandleRecordings handle requests to /recordings/ endpoint
+func (s *LivepeerServer) HandleRecordings(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		w.WriteHeader(http.StatusMethodNotAllowed)
```
Add error logs for failure paths?

darkdarkdragon replied:
Fixed.

yondonfu reviewed (changes requested):
I see that we're rolling over JSON playlists every hour and each JSON playlist is identified using a timestamp. So, the external OS ends up with a bunch of those JSON playlists and then we fetch all the playlists prefixed with a specific manifestID to combine them together in HandleRecordings?

darkdarkdragon commented:
> I see that we're rolling over JSON playlists every hour and each JSON playlist is identified using a timestamp. So, the external OS ends up with a bunch of those JSON playlists and then we fetch all the playlists prefixed with a specific manifestID to combine them together in HandleRecordings?

Yes, exactly.

darkdarkdragon committed:
Error logging in HTTP push handler

darkdarkdragon committed:
move save retry logic to the separate function

yondonfu commented on drivers/drivers.go:
```diff
@@ -175,6 +175,22 @@ func ParseOSURL(input string, useFullAPI bool) (OSDriver, error) {
+// SaveRetried tries to SaveData specified number of times
+func SaveRetried(sess OSSession, name string, data []byte, meta map[string]string, retryCount int) (string, error) {
+	if retryCount < 1 {
+		return "", fmt.Errorf("Invalid retry cound %d", retryCount)
```
"cound" -> "count"

darkdarkdragon replied:
fixed

yondonfu reviewed (approved):
Left a comment about a typo but feel free to fix during the squash.

Besides that LGTM
